### PR TITLE
Use RBAC and bicep provisioning for Azure OpenAI

### LINKED ISSUE #2490: Add Azure Open AI provisioning bicep
We have Azure AI Search bicep / provisioning:

https://github.com/dotnet/aspire/blob/420962e613f20b63ea86829fe8f55dc591e8bac7/src/Aspire.Hosting.Azure/Bicep/search.bicep#L47-L59

We should do the same for Azure Open AI.

Note that both should use RBAC by default. See https://github.com/dotnet/aspire/issues/2444.

cc @sebastienros

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #2490
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2529)

### CODE REVIEW

**Comment by sebastienros** [comment]
Couldn't test E2E right now as it blocks on quotas. Won't merge before this is resolved.

File: src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:42]: Do we want a context object for these arguments.

File: tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs
Comment (eerhardt) [inline tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs:377]: Can you change this test to verify the manifest like the following?

https://github.com/dotnet/aspire/blob/525c59ad9d58de54c71071e248b060061dfecd56/tests/Aspire.Hosting.Tests/SqlServer/AddSqlServerTests.cs#L123-L153

File: src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:42]:
```suggestion
    public static IResourceBuilder<AzureOpenAIDeploymentResource> AddDeployment(this IResourceBuilder<AzureOpenAIResource> builder, string name, string modelName, string modelVersion, string skuName = "Standard", int skuCapacity = 2)
```

What does `serverBuilder` mean here?

File: src/Aspire.Hosting.Azure/Bicep/openai.bicep
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Bicep/openai.bicep:48]: Why batchSize 1?

File: src/Aspire.Hosting.Azure/AzureOpenAIDeploymentResource.cs
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/AzureOpenAIDeploymentResource.cs:18]: We should fill these out.

File: playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/Program.cs
Comment (davidfowl) [inline playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/Program.cs:9]: No fluent API?

File: src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:45]: Do these deployments need to be a "top level" resource? That means their name must be unique across the whole app, which I'm not sure is necessary as they don't need to be referenced by the app (i.e. they don't have a connection string or similar).

File: playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/Program.cs
Comment (sebastienros) [reply playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/Program.cs:9]: Do you know of a resource I should use as an example?

File: src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs
Comment (sebastienros) [reply src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:45]: I don't think they need to be either. Does it mean remove the `AzureOpenAIDeploymentResource` class?

Comment (eerhardt) [reply src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:45]: I think you still need to have an "AzureOpenAIDeployment" class of some sort. I just question if it needs to be a full fledged "Resource". @davidfowl and @mitchdenny should weigh in here.

File: src/Aspire.Hosting.Azure/Bicep/openai.bicep
Comment (sebastienros) [reply src/Aspire.Hosting.Azure/Bicep/openai.bicep:48]: This ensures that these are deployed sequentially as by default they are done in parallel. This was in the original template from @jongio so I assume this was recommended at least.

File: src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Extensions/AzureOpenAIExtensions.cs:45]: These don't need to be a resource. This just seems like input to the original resource. This is what we did with service bus and cosmos.

File: src/Aspire.Hosting.Azure/Bicep/openai.bicep
Comment (jongio) [reply src/Aspire.Hosting.Azure/Bicep/openai.bicep:48]: You'll run into concurrency issues without it - this makes them serial.

**Comment by davidfowl** [comment]
Looks good.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting.Azure/AzureOpenAIDeployment.cs
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/AzureOpenAIDeployment.cs:14]: Is `Standard` and `2` applicable for all deployment models?

**Review by eerhardt — APPROVED** [review]
LGTM. Nice work!

File: playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/aspire.hosting.azure.bicep.openai.bicep
Comment (eerhardt) [inline playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/aspire.hosting.azure.bicep.openai.bicep:75]: Why is this duplicated now?

Comment (sebastienros) [reply playground/OpenAIEndToEnd/OpenAIEndToEnd.AppHost/aspire.hosting.azure.bicep.openai.bicep:75]: Snap, mean automerger ... It got generated with the tool, not sure why. When looking at the diff locally I thought it was just adding a new line. Will fix it.

**Pull request merged**
